### PR TITLE
feat(gms/sglang): persist native unified-cache pages

### PR DESCRIPTION
**Pull request by hutm**
Part 4 of 14 in #14704. Builds on #14708.

## Why

Surviving bytes are not enough. A replacement also needs the content-hash to native-page mapping for completed prefixes.

## What changes

- Provide a GMS-aware subclass of SGLang's native UnifiedRadixCache.
- Publish only completed immutable prefixes.
- Rebuild SGLang's native radix-tree view by adopting validated pages.
- Keep lookup and eviction policy inside SGLang instead of duplicating its tree in GMS.

This adds the cache adapter and metadata flow, not failover orchestration.

## Validation

Native hash compatibility, page ordering, publication, adoption, and eviction tests pass at this prefix.

Next: #14710 makes the transactions fail closed.

### CODE REVIEW

File: lib/gpu_memory_service/integrations/sglang/model_loader.py [hunk 1]
```diff
@@ -105,43 +123,72 @@ def _load_import_only(self, model_config, device_config, impl) -> torch.nn.Modul
...
+        # run only the structural post-load hook before GMS binds the final
+        # tensors published by the writer.
+        with set_default_torch_dtype(model_config.dtype):
+            with meta_device:
+                model = _initialize_model(
+                    model_config,
+                    load_config,
+                    quant_config,
+                )
+                _post_load_weights(model)
```
[the remark this answers is not in the document]
Comment (hutm) [reply lib/gpu_memory_service/integrations/sglang/model_loader.py]: Addressed in 2e80260abe2. We no longer run arbitrary `post_load_weights` code on meta tensors—or replay it against shared published weights: custom post-load models fail closed before materialization, while the one known runtime-only FlashInfer state is restored explicitly after `materialize_module_from_gms`; focused model-loader tests cover both decisions.

File: lib/gpu_memory_service/integrations/sglang/patches.py [hunk 2]
```diff
@@ -185,23 +188,345 @@ def patched_alloc_memory_pool(self, *args, **kwargs):
...
+        )
+        return
+    if getattr(Scheduler, "_gms_serving_timeout_patched", False):
+        _serving_timeout_patched = True
+        return
+
+    original_run_event_loop = Scheduler.run_event_loop
+
+    def patched_run_event_loop(self, *args, **kwargs):
+        # First (and only) entry == post-warmup, pre-traffic: tighten now.
```
[the remark this answers is not in the document]
Comment (hutm) [reply lib/gpu_memory_service/integrations/sglang/patches.py]: Addressed in 2e80260abe2: removed the redundant lifecycle comment and retained the detailed constraint in `patch_serving_collective_timeout_for_gms`’s docstring.

File: lib/gpu_memory_service/integrations/sglang/patches.py [hunk 3]
```diff
@@ -185,23 +188,344 @@ def patched_alloc_memory_pool(self, *args, **kwargs):
 
         return original_alloc_memory_pool(self, *args, **kwargs)
 
-    ModelRunner.alloc_memory_pool = patched_alloc_memory_pool
+    # New SGLang versions ask the loader for ``preloaded_weights_bytes`` and
+    # apply this correction natively. Keep the wrapper only for older versions;
+    # applying both would double-count imported GMS weights and oversize KV.
+    if not hasattr(ModelRunner, "account_preloaded_weights"):
```
[the remark this answers is not in the document]
Comment (hutm) [reply lib/gpu_memory_service/integrations/sglang/patches.py]: No code change: current SGLang already calls account_preloaded_weights immediately before alloc_memory_pool in scheduler.py init_target_memory_pool. Applying the adjustment again in the GMS adapter would double-count weights on the pinned runtime.

File: lib/gpu_memory_service/integrations/sglang/gms_unified_cache.py [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+            self, req, is_insert: bool = True, *, kv_len_to_handle: int, **kwargs
+        ) -> None:
+            from sglang.srt.mem_cache.radix_cache import RadixKey
+
+            token_ids = (req.origin_input_ids + req.output_ids)[:kv_len_to_handle]
+            key = RadixKey(
+                token_ids,
+                req.extra_key,
+                is_bigram=self.tree_core.is_eagle,
+                cache_salt=req.cache_salt,
```
[the remark this answers is not in the document]
Comment (hutm) [reply lib/gpu_memory_service/integrations/sglang/gms_unified_cache.py:145]: No code change: this finding does not match the exact pinned/current SGLang runtime used by the stack (fbf8f1dbf69f). Its RadixKey constructor includes cache_salt, and the real-runtime TP1/TP2 tests exercise this completion path successfully. Removing it would instead collapse distinct native cache identities.

File: lib/gpu_memory_service/integrations/sglang/gms_unified_cache.py [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+            return hashes
+
+        def _publish_finished_prefix(self, key) -> None:
+            from sglang.srt.mem_cache.base_prefix_cache import MatchPrefixParams
+
+            result = super().match_prefix(MatchPrefixParams(key=key))
+            indices = result.device_indices
+            if len(indices) != len(key):
+                raise RuntimeError("completed SGLang prefix is not fully resident")
+            leases = retain_hbm_indices(self.token_to_kv_pool_allocator, indices)
```
[the remark this answers is not in the document]
Comment (hutm) [reply lib/gpu_memory_service/integrations/sglang/gms_unified_cache.py:98]: Addressed before activation by d24561433df (#14710): the backend is accepted only in authoritative mode, and directory-pressure reclamation removes visibility generation-conditionally before releasing retained pages. Shadow mode now fails closed.

File: lib/gpu_memory_service/integrations/sglang/install_gms_unified_cache.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
+# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+"""Register GMS as an SGLang UnifiedRadixCache backend."""
```
[the remark this answers is not in the document]
Comment (hutm) [reply lib/gpu_memory_service/integrations/sglang/install_gms_unified_cache.py]: Fixed in 3f1ffa49486; the redundant module docstring is removed.

File: lib/gpu_memory_service/integrations/sglang/gms_unified_cache.py [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+            # TreeCore normally computes these hashes only for HiCache or an
+            # external linker. GMS needs the same native hashes, but neither
+            # feature. This flag is TreeCore-local and only controls hashing.
+            self.tree_core.enable_storage = True
+            self.tree_core.backfill_missing_hash_values()
+            self._gms_directory.start_async_read()
+
+        @staticmethod
+        def _hashes_for_key(key, page_size: int) -> list[str]:
+            hashes = get_hash_str(key, page_size=page_size)
```
[the remark this answers is not in the document]
Comment (hutm) [reply lib/gpu_memory_service/integrations/sglang/gms_unified_cache.py:87]: Addressed in the immediately following correctness PR, d24561433df (#14710), before runtime activation in #14711. Directory identity there uses SGLang's native page hash plus canonical extra_key/cache_salt namespace bytes, preventing cross-input aliasing.

File: lib/gpu_memory_service/integrations/sglang/install_gms_unified_cache.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        raise RuntimeError(f"SGLang cache backend {_BACKEND!r} is already registered")
+    register_radix_cache_backend(_BACKEND, _factory)
+    return True
+
+
+def configure(server_args) -> bool:
+    if not _enabled():
+        return False
+    install()
+    from sglang.srt.arg_groups.overrides import declare_late_resolution, resolving_view
```
[the remark this answers is not in the document]
Comment (hutm) [reply lib/gpu_memory_service/integrations/sglang/install_gms_unified_cache.py]: Fixed in 6f8783177db.  now discovers the resolution APIs with , uses raw  when  is absent, and follows the same /direct-assignment fallback used by . The legacy path has focused regression coverage.

File: lib/gpu_memory_service/integrations/sglang/install_gms_unified_cache.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+
+def _enabled() -> bool:
+    return (
+        os.environ.get("GMS_KV_DIRECTORY_MODE", "off").lower()
+        in (
+            "shadow",
+            "authoritative",
+        )
+        or os.environ.get("GMS_SGLANG_ENABLE_KV_RING") == "1"
```
[the remark this answers is not in the document]
Comment (hutm) [reply lib/gpu_memory_service/integrations/sglang/install_gms_unified_cache.py:21]: Addressed in #14710 before runtime activation: `_enabled()` now requires `GMS_KV_DIRECTORY_MODE=authoritative`; ring-only mode gates allocation but cannot select the directory-dependent cache backend.
